File: go/launcher/proxy/screenshot/screenshot.go
// Copyright 2016 Google Inc.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

// Package screenshot includes a handler for the WebDriver screenshot operation.
package screenshot

import (
	"bytes"
	"context"
	"encoding/base64"
	"fmt"
	"image"
	"image/png"
	"net/http"

	"github.com/bazelbuild/rules_webtesting/go/cropper"
	"github.com/bazelbuild/rules_webtesting/go/launcher/proxy/driverhub"
	"github.com/bazelbuild/rules_webtesting/go/launcher/proxy/webdriver"
)

const sizeScript = `return {"width": screen.width, "height": screen.height};`

// ProviderFunc provides a handler for /screenshot command that crops the image if Chrome mobile emulation is enabled.
func ProviderFunc(session *driverhub.WebDriverSession, desired map[string]interface{}, base driverhub.HandlerFunc) (driverhub.HandlerFunc, bool) {
	if !MobileEmulationEnabled(desired) {
		return base, false
	}

	return func(ctx context.Context, rq driverhub.Request) (driverhub.Response, error) {
		if rq.Method != http.MethodGet || len(rq.Path) != 1 || rq.Path[0] != "screenshot" {
			return base(ctx, rq)
		}

		img, err := GetMobileScreenshot(ctx, session.WebDriver)
		if err != nil {
			return driverhub.ResponseFromError(err)
		}

		return CreateResponse(img)
	}, true
}

// MobileEmulationEnabled determines if the capabilities define a mobile emulate config.
func MobileEmulationEnabled(caps map[string]interface{}) bool {
	chromeOptions, ok := caps["chromeOptions"].(map[string]interface{})
	if !ok {
		return false
	}

	if _, ok := chromeOptions["mobileEmulation"]; !ok {
		return false
	}

	return true
}

// GetMobileScreenshot crops a screenshot based on the emulated viewport size.
func GetMobileScreenshot(ctx context.Context, driver webdriver.WebDriver) (image.Image, error) {
	img, err := driver.Screenshot(ctx)
	if err != nil {
		return nil, err
	}

	val := struct {
		Width  int
		Height int
	}{}

	if err := driver.ExecuteScript(ctx, sizeScript, nil, &val); err != nil {
		return nil, err
	}

	return cropper.Crop(img, image.Rect(0, 0, val.Width, val.Height))
}

// CreateResponse creates a /screenshot endpoint response for a given image.
func CreateResponse(img image.Image) (driverhub.Response, error) {
	buffer := &bytes.Buffer{}
	b64 := base64.NewEncoder(base64.StdEncoding, buffer)

	if err := png.Encode(b64, img); err != nil {
		return driverhub.ResponseFromError(err)
	}

	b64.Close()

	return driverhub.Response{
		Status: http.StatusOK,
		Body:   []byte(fmt.Sprintf(`{"status": 0, "value": %q}`, buffer.String())),
	}, nil
}

File: java/com/google/testing/web/screenshotter/Screenshot.java
// Copyright 2017 Google Inc.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//
// //////////////////////////////////////////////////////////////////////////////
//
package com.google.testing.web.screenshotter;

import com.google.errorprone.annotations.Immutable;
import java.awt.image.BufferedImage;
import java.io.ByteArrayInputStream;
import java.io.IOException;
import java.util.Base64;
import javax.imageio.ImageIO;
import javax.imageio.ImageReader;
import org.json.JSONException;
import org.json.JSONObject;

/** A wrapper around a screenshot returned by Screenshotter. */
@Immutable
public final class Screenshot {
  private static final String FORMAT = "png";

  private final String base64;

  Screenshot(JSONObject response) throws JSONException {
    this.base64 = response.getString("value");
  }

  /** Returns the BASE 64 string returned by the screenshot endpoint. */
  public String asBase64() {
    return base64;
  }

  /**
   * Returns the decoded bytes for the screenshot. Every call to this returns a new copy of the
   * decoded bytes.
   */
  public byte[] asBytes() {
    return Base64.getDecoder().decode(base64);
  }

  /**
   * Returns a BufferedImage of the screenshot. Every call to this returns a new copy of the image.
   */
  public BufferedImage asImage() throws IOException {
    ImageReader imageReader = ImageIO.getImageReadersByFormatName(FORMAT).next();
    imageReader.setInput(ImageIO.createImageInputStream(new ByteArrayInputStream(asBytes())), true);
    return imageReader.read(0);
  }

  @Override
  public boolean equals(Object other) {
    if (other == null) {
      return false;
    }
    if (!(other instanceof Screenshot)) {
      return false;
    }
    Screenshot o = (Screenshot) other;
    return asBase64().equals(o.asBase64());
  }

  @Override
  public int hashCode() {
    return asBase64().hashCode();
  }
}
